### DISCUSSION TITLE
[RFC] ESLint out of the box

### DISCUSSION THREAD

**Discussion by prateekbh** [vercel/next.js#17320] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.
Currently ESLint is the fourth most installed dependency in Next.js apps. The entire JavaScript ecosystem heavily relies on `eslint` to avoid obvious errors, anti practices, and [performance gotchas](https://www.npmjs.com/package/@next/eslint-plugin-next) in case of Next.js.
Due to this, it makes sense to provide this functionality out of the box, automatically configured for all Next.js apps with zero effort from developers. This feature aims to bring ESLint configured for all new and existing apps as a part of the Next.js `build` and `dev` command. 
Next.js apps will be linted against the configured eslint configuration and the build will log all the errors and the warnings as a result of the linting.

**Note:** This will also enable the `eslint-next-plugin` for all Next.js apps giving them insight not only about anti-patterns in the JavaScript/React land but also save them from performance footguns arising from misusing the Next.js primitives.

## Describe the solution you'd like

We’d like to introduce a custom eslint loader in the webpack toolchain. This loader will enable us to lint the first party code with the configured `.eslintrc.*` file, and allow Next.js to surface errors/warnings as discovered by eslint. 
A loader will also allow us to not only lint the entire codebase during `next build` but also during `next dev`. Additionally it will re-lint only the changed files at the time of the dev server.

The currently well established `eslint-loader` and `babel-loader` are implemented to work with text content. They do not share any parsing work which increases the build time cost for a compilation. Our custom loader intends to share the AST built from the textual content in `eslint-loader` and pass it to the `next-babel-loader` for it to consume it directly.

Some of the differences from the currently well established eslint-loader would be as follows
- The eslint errors will be added back to the webpack compilation as compilation errors. This allows us to expose the list of all lint errors/warnings before breaking the build.
- The parser used for eslint will be `babel-eslint` instead of the default `espree`. This is inline with the rest of the toolchain to be centered around a single parser/compiler.
- The AST produced by the above parser in `eslint-loader` will be forwarded to the `babel-loader`. The `babel-loader` can skip the parsing of string content to AST and directly work on the AST provided to it.

If a base `.eslintrc.*` file is found, the webpack loader would consume it. But if no such qualifying file is present then the loader would work with a default config (`eslint:recommended`, `eslint-plugin-react-hooks:recommended` and `eslint-plugin-next:recommended`).

## Describe alternatives you've considered

Alternatively we can run a parallel eslint along with the build and the dev server but this would require additional efforts of maintaining a message channel between the linter and the main compilation and a separate process management overhead.
This would also mean that there would be overhead of double parsing of the code and thus more CPU would be required than if they were to share some of the work and happen sequentially.

## Additional context

As a part of this change, all Next.js apps will be linted with the help of [eslint-plugin-next](https://www.npmjs.com/package/@next/eslint-plugin-next) in addition to the config found in `eslintrc.*`. This is an important step in making sure that Next.js is configured correctly and its components are being used in the most efficient way.

For apps to opt-out of this, they will need to opt-out of the individual rules and will not be able to
opt-out of the entire plugin at once. This will be made this way to ensure that the opt-outs are always intentional.

Also, `.eslintignore` will not be supported and all first party files that are going to be bundled by NextJS will be linted by this system. Any file or line inclusions should be done by the help of inline eslint commands.

**Comment by eps1lon** [comment]
How will the perf impact on `next build` look like? We currently have a separate CI pipeline (circleCI) with spare resources and a deploy pipeline (netlify) that is already starving. `next build` should really only be concerned with building what's necessary for the deploy and not do any static analysis.

We can already [disable type-checking when building](https://nextjs.org/docs/api-reference/next.config.js/ignoring-typescript-errors) so the same should be possible for additional static analysis.

> A loader will also allow us to not only lint the entire codebase during next build but also during next dev. 

A codebase using nextjs does not only consist of `next` related files. It can contain auxiliary scripts or an entire (published) component library. If one already has a dedicated lint task then it'll now be split up into the existing lint task and `next build`.

Having a dedicated plugin for perf gotchas can definitely be helpful but one should be able to integrate it into their existing lint task and opt-out of the `next build` integration.

**Reply by prateekbh** [reply]
Thank you for the detailed feedback @eps1lon! This is very helpful.

The suggestion to have a blanket opt-out is definitely be something we would seriously take into consideration. I do see a different use case of prototyping and beginner's learning experience to use such an opt-out. However, we would strongly want all production apps to use this and make sure that their builds/dev server(in this or later version) would break for anything which regresses performance.
As per other concerns I can address below.

> How will the perf impact on next build look like?

The feature will definitely not be free of cost. However what differentiate this from type-checking is that this is a deliberately kept in webpack toolchain. This allows the loader to share work with other parts like babel-loader e.g. AST parsing. This was already an intentional design decision to address the cost on `build`

>A codebase using nextjs does not only consist of next related files. 

I feel this is already a step towards similar thinking. If a repo has other stuff(a design system/ backend API etc) than just a Next.js app then other parts might need a different lint setup as compared to a Next.js production app. So the integration with `build` can take additional care that the app is atleast linted with the nextjs lint plugin. However if the app happens to have a `eslintrc.*` file, it will also lint it with the rules defined inside it.

**Comment by dane-brown** [comment]
I think this would be a really beneficial change, sometimes it can be a struggle to implement eslint in a project with reliable rules and by this coming out of the box it encourages good standards. After reading the RFC I do think it could be important to be able to disable eslint entirely as I don’t think we should assume everyone will want to use it, although good standards are important we can’t assume everyone will want to use it and shouldn’t be forced to do so.

**Comment by m1guelpf** [comment]
How would this affect people who lint on save with VSCode or other code editors?

**Comment by prateekbh** [comment]
@m1guelpf 

> How would this affect people who lint on save with VSCode or other code editors?

This particular feature integrates the capability of linting with build. But it will not affect any other capability of other integrations. Since we still consume on `eslintrc.*`. The other integrations like code editors would still be unaffected. The editors and other integrations will still continue to help you see and fix those errors.

**Comment by juliuscc** [comment]
**TLDR:** I think it is a (kinda*) good idea to run eslint during build but definetly not during dev.

### 👎  Blocking my dev workflow
One of my largest pain points with CRA is that it runs eslint on the dev stage. If I comment out some piece of code I often get an error about unused variables and then I have to comment out the variables. Then I have to comment out the variables used to create the commented variables. And so on.

When I am done fixing eslint errors I have lost track of what I was changing. The point is that during dev you often have code that should not pass a lint, but should be shown on the screen. I strongly dissagree with the idea of running linting during dev.

In Nexts case I really like that type checking is disabled during dev, but enabled during build. That means that I can explore, and test unfinished concepts during dev very quickly. At the same time I get safety provided by type checking when I build. If eslint would be introduced for Next I think the same kind of workflow should be best.

### 👎 Slowing down my dev workflow
I also hate how eslint makes building slower during dev. During dev it is extremely important to see changes quickly. That's why I use Next instead of CRA. Every extra milisecond increases the probability of the developer to loose focus and forget context. This problem is less of an issue during build as I don't care if it takes 5 seconds or 15 seconds.

### 👍 Easier config
I like the idea of having a sensible eslint-config included by default. Especially if it adds next-specific rules. I am however conflicted about issues like what to do if teams want a completely different eslint-config. I personally like using a11y-rules, but others might not. Should that be included or not. Why shouldn't I be able to add/remove it to/from the default if I don't agree with the default.

I think adding a command like `next lint` would be a great middle ground to test this feature out. That way we can find the pain points before making it compulsary. And you still get a nice eslint-config by default with next. I would definetly use it for small projects, where I don't care about specific rules.

**Reply by prateekbh** [reply]
> I am however conflicted about issues like what to do if teams want a completely different eslint-config.

Will it be ok if the eslint config is totally configurable and upto the app developer but the next specific rules are always applied no matter what?

**Reply by prateekbh** [reply]
Around the lint in dev setup: leaving the time concern apart, Will it help if the dev is never broken but the warnings/errors are shown to you in an overlay in your browser?

**Reply by juliuscc** [reply]
I like both of those suggestions a lot. For the next-specific rules it is good if they are lean but that seems to be the case. For the overlay it is important to me that I can ignore it if I want. In other words it should not be a full screen dialog but a small indicator in the corner that I can press for more info.

**Comment by MathiasKandelborg** [comment]
I highly enjoy ESLint and TypeScript, but I do not think that Next should touch ESLint in any way. ESLint and Next are two different kinds of software that each do their thing very, very well. I personally can't live without it!

My first and immediate concern is [SoC](https://en.wikipedia.org/wiki/Separation_of_concerns). TypeScript and Babel - arguably - makes more sense to integrate with a framework that builds and minifies an application. But having static analysis forced upon me, without the ability to configure AND no .eslintignore support, while possibly slowing down my workflow is not something I like. I even think it would break my current setup (using ESLint via Yarn Berry in another folder, not in a node_modules dir).

It is exceedingly easy to integrate eslint and the next plugin with a project. How about keeping SoC and just provide an ESLint configuration + installation in every example and with create-next-app? That way, everyone is getting it by default, unless they're starting from absolute scratch, where one _should_ imagine the user knows what they want!

By forcing ESLint on everybody, without exception, as the base solution; you're telling everybody they're not good enough, at least not good enough to make their own choices. There should _always_ be the ability to opt entirely out, that is, not use ESLint at all.


It is against SoC, bloat to the Next.js package, forces a specific linting package on people, and I think not supporting `.eslintignore` is just wrong.


If this is a route to take, please just use the actual software, install that for the user, provide a default configuration, show them the ESLint website, and tell them where to start. I don't like abstracting things like a whole ecosystem into Next.js; it's looking more like Gatsby's stuff, assuming everyone sucks at configuring their project and doesn't want to learn something new/explore. At the very least, make it optional and not something next includes in its package and build system.

**Reply by rauchg** [reply]
Strongly disagree with this. ESLint is **essential** for shipping React to production, especially with Hooks relying on linting for correctness and guidance. 

Not having it and hoping that people remember to install it and configure it is reminiscent of the problems Next.js set out to solve in the first place, back when people still used to ship React in dev mode to prod.

Furthermore, it's very much along the same lines of our TypeScript integration, which is extremely popular with Next.js users. Next.js is much more than just a "zero-config bundler": it's a comprehensive **solution** for shipping high-performance pages without bugs and regressions!

**Reply by MathiasKandelborg** [reply]
I want to stress that I am saying I would love default ESLint support, but shoving it inside Next.js and providing no `.eslintignore` support is not something I like. 

It might have come off a bit harsh, and I was not very clear on my point, but I stand by my opinion on SoC. Let me try and clarify:

I do not think ESLint should be handled internally by Next.js. The maintainability cost could be big given the different approaches with which people are using ESLint; some run eslint _programmatically_, while others:
- On saving each file
- With different rulesets in dev and before build
- On commit
- Globally

Here is what I think could be **the solution**:

If ESLint is enabled in the next config (a proposed setting), which would be by default, Next.js provides helpers to show the errors in the overlay or console or both! That could be very nice to have. I do not know how hard tapping into ESLint is, but it should be doable.

Provide the ESLint package in all examples, and ship it as a dependency with create-next-app along with a default configuration.

Another addition could be to have the packages as peer dependencies to prompt people to install the needed packages.

The main point I am trying to make is that Next.js should not prefer _how_ or _when_ ESLint is used. I do, however, believe it would be great to have ESLint enabled and configured, **by default**, every time someone uses create-next-app or clone an example. While it would be 'easier' to handle ESLint internally, I think that approach is going to be restrictive for some workflows and provide unwanted behavior in edge-cases.

Furthermore, my remark on "Babel and TS - arguably -makes more sense" was not useful for much. I agree they should be a part of Next.js, as it is synonymous with integrating those oneself. 
I do not think that TypeScript, Webpack (which I did not mention before but is relevant) nor Babel is similar to ESLint; their concerns lie in a different area. TypeScript is a superset of JS to provide typing, Babel transpiles code to web-safe code, and Webpack packs all the things (haha).
They are all manipulating the project in one way or another, to **enable** the finished Next.js application.
ESLint maybe (usually) enforces a code style and provides a way to lint a JS/TS codebase, but it does not change the code in any way (unless you apply the fix), which I think is a somewhat big difference in concern.


Thinking of Next.js as a comprehensive solution to ship without bugs and regressions, ESLint is essential because there are no alternatives - that I know of - and the tooling it provides is essential to avoid common performance pitfalls/bugs.

I really think there is a place for ESLint with Next.js _every single time,_ but it should be included as-is, not within Next.js. It should not be forced at any given point, and be entirely opt-out and configurable (via an ignore file). That way, no one is left behind... for good.

**Reply by juliuscc** [reply]
Just to reinforce @prateekbh's point. I _always_ run eslint as a test step on the ci-server, but _never_ while developing locally. It is for the same reason as I do not run tests for every local dev refresh. It makes sense to get a warning when I am done. If used during dev it adds latency when working on the web page.

**Reply by eps1lon** [reply]
This is basically the same concern as https://github.com/vercel/next.js/discussions/17320#discussioncomment-84162: Linting and type-checking as zero-config is helpful. But duplicating that step if I already have it setup is not good developer experience. 

It wouldn't matter if CI/CD ressources where free but they aren't. If nextjs decides to not build in a switch for linting I have to monkey-patch it which results in an overall worse product.

What are the arguments to allow disabling type-checking but not allow disabling linting?

**Comment by rauchg** [comment]
To further emphasize how essential linting is to the React ecosystem today, I Google'd "Rules of Hooks" and the second sentence already mentions linting 

![6563EEBB-F39A-4DC7-8844-C8D2BF56E3D6](https://user-images.githubusercontent.com/13041/95263768-2d112100-07e3-11eb-9561-52b9639f9f28.jpeg)

This is the kind of stuff that I never want to set up manually (or worse: forget to set up altogether!) and why I reach for Next.js every time to figure this out for me 💘

**Reply by MathiasKandelborg** [reply]
I think this further proves my point with SoC. React doesnt handle the linting, it leaves that to the user.

This points to how essential it truly is. I think Next should provide a default config and perhaps even require ESLint. Not incorporate it in a specific way, handle it internally, and without an opt out option.
